### PR TITLE
Port DOS protections to our protojson fork

### PR DESCRIPTION
**Pull request by tdeebswihart**
**What changed?**

I ported over the [DOS fixes](https://github.com/protocolbuffers/protobuf-go/commit/bfcd6476a38e41247d6bb43dc8f00b23ec9fffc2) from the official protobuf repository

**Why?**

Two DOS bugs were fixed by golang/protobuf recently:

1. https://github.com/golang/protobuf/issues/1583
2. https://github.com/golang/protobuf/issues/1584

**How did you test it?**

I ported over their tests for the issue

**Potential risks**

None,

### CODE REVIEW

**Review by cretz — APPROVED** [review]
In a perfect world we'd maintain just a set of patches on top of existing lib, but that is annoying.

File: internal/protojson/well_known_types.go [hunk 1]
```diff
@@ -309,48 +309,25 @@ Loop:
...
-				if err := d.skipJSONValue(); err != nil {
-					return err
-				}
+		switch tok.Kind() {
+		case json.ObjectClose, json.ArrayClose:
+			open--
+		case json.ObjectOpen, json.ArrayOpen:
+			open++
+			if open > d.opts.RecursionLimit {
+				return errors.New("exceeded max recursion depth")
```
Comment (rodrigozhou) [inline internal/protojson/well_known_types.go:324]: You're not doing recursion anymore, so does it matter to track the depth?

Comment (tdeebswihart) [reply internal/protojson/well_known_types.go:324]: This is copied verbatim from the official codebase

**Comment by tdeebswihart** [comment]
> In a perfect world we'd maintain just a set of patches on top of existing lib, but that is annoying.

We'd need to fork the whole protobuf stack to do so, which would suck. We could do that and periodically rebase our changes on the upstream, but that's a lot of work

**Pull request merged**